### PR TITLE
Fix more monthly bugs and add more tests

### PR DESCRIPTION
**Pull request by iterion**
I discovered a few more edge cases. I took some time to refactor the `handle_monthly` function a bit and it's now a bit easier to reason about. ~But, I think there is probably more that can be done to simplify it.~ I went as far as I could to simplify it and keep the tests running. Along the way, I noticed an incorrect offset in an assertion that the new code doesn't generate.

### CODE REVIEW

**Comment by shavz** [comment]
`by_month` was the first function I implemented so it was a bit rusty, thank you for your time and effort - reviewing it now

File: src/lib.rs [hunk 1]
```diff
@@ -427,83 +427,25 @@ impl<'a> RRule<'a> {
...
-                            }
-                        }
-                    }
-                }
-                if next_date.month().eq(&(by_month.parse::<u32>().unwrap())) {
-                    break;
-                }
+        if !by_month.is_empty() {
+            let by_month_u32 = by_month.parse::<u32>().unwrap();
+            if by_month_u32 <= next_date.month() {
```
Comment (shavz) [inline src/lib.rs]: Not sure about this bit,

so if the current month is 6 and BYMONTH=6, I would still like the next date to fall in this current month and not move a year forward?

Comment (iterion) [reply src/lib.rs]: Good catch! It seems we didn't have any test cases on this edge case. I started refactoring daily and was able to reuse a lot of this same implementation. Due to that, I noticed that this particular conversion fell short. But, I think I have a fix that will work for both daily and monthly. I'll be sure to add new test cases to cover.

**Comment by iterion** [comment]
@shavz any additional concerns on this PR? I have another PR for refactoring daily pending that's based on this branch that I'd like to send over for your consideration, just waiting for this to merge so I can rebase and present a clean diff.

**Comment by shavz** [comment]
apologies, life got in the way, no concerns! 

Thanks for simplifying the logic flow.
